src/crowdsec.ts

```diff
@@ -0,0 +1,344 @@
+    let ipRangesChanged = false
+
+    if ("deleted" in respJSON && Array.isArray(respJSON["deleted"])) {
+        respJSON["deleted"].forEach((decision: any) => {
+            decision = normalizeDecision(decision)
+            if(decision.scope == "range"){
+                delete currentIPRanges[decision.value]
+                ipRangesChanged = true
+            } else {
+                ngx.shared.crowdsec_decision_store.delete(decision.value)
+            }
+        })
+    }
+
+    if ("new" in respJSON && Array.isArray(respJSON["new"])) {
+        respJSON["new"].forEach((decision: any) => {
+            decision = normalizeDecision(decision)
+            if(decision.scope == "range"){
+                currentIPRanges[decision.value] = decision.type
+                ipRangesChanged = true
+            } else {
+                ngx.shared.crowdsec_decision_store.set(decision.value, decision.type.toString())
+            }
+        })
+    }
+    if(ipRangesChanged){
+        ngx.shared.crowdsec_decision_store.set("IP_RANGES", JSON.stringify(currentIPRanges))
+    }
+}
+
+function normalizeDecision(decision: any): any {
+    decision.value = decision.value.toLowerCase()
+    decision.type = decision.type.toLowerCase()
+    decision.scope = decision.scope.toLowerCase()
+    decision.origin = decision.origin.toLowerCase() 
+    return decision
+}
+
+function serveBanTemplate(r: NginxHTTPRequest): void {
+    r.headersOut["Content-Type"] = "text/html"
+    if ("ban_template_path" in r) {
+        return fs.readFileSync(r.ban_template_path)
+    }
+    return fs.readFileSync(config.ban.template_path)
+}
+
+function isCaptchaSubmission(r: NginxHTTPRequest): boolean {
+    return "captcha_token" in r.args
+}
+
+async function captchaSubmissionIsCorrect(r: NginxHTTPRequest): Promise<boolean>{
+    let siteVerifyURL:string
+
+    if(config.captcha.provider == "google_recaptcha_v2"){
+        siteVerifyURL = "https://www.google.com/recaptcha/api/siteverify"
+    } else if(config.captcha.provider == "hcaptcha"){
+        siteVerifyURL = "https://hcaptcha.com/siteverify"
+    } else {
+        siteVerifyURL = "https://challenges.cloudflare.com/turnstile/v0/siteverify"
+    }
+
+    const captchaToken = r.args["captcha_token"]
+    const body = querystring.stringify({
+        secret:  config.captcha.secret_key,
+        response: captchaToken,
+        remoteip: r.remoteAddress
+    });
+    const result =  await ngx.fetch(siteVerifyURL, {
+        body: body,
+        method: 'POST',
+        headers: {
+        'Content-Type': 'application/x-www-form-urlencoded'
+        }
+    });
+    
+    const outcome =  await result.json();
+    return outcome["success"] === true;
+}
+
+
+async function serveCaptchaSubmissionHandler(r: NginxHTTPRequest): Promise<void> {
+    if(await captchaSubmissionIsCorrect(r) == true){
+        // @ts-ignore
+        const cookie = jwt.encode({"exp": Date.now()/1000 + 1800}, config.captcha.secret_key + r.remoteAddress)
```

Comment:
JWT are not encrypted so you are potentially leaking the captcha secret key here. Put an example key in https://jwt.io/ you can see its not encrypted

Reply:
You can site_key instead of secret key as that is client facing code

Reply:
Actually I misread the code, you using that as the signing key